### PR TITLE
[Service Bus] Live Tests: Stabilization and Resilience (Phase 2)

### PR DESCRIPTION
**Pull request by jsquire**
## Summary

Some of the Service Bus client library Live tests have a measure of non-determinism in them due to timing issues or transient network failures.  Because of the nature of message passing send/receive patterns, many of these tests have non-trivial logic around timing and retries for service interactions which can make them unreliable but tricky to troubleshoot and stabilize.

In their current form, these tests are each passing at roughly a 98% rate after isolation with intermittent failures and no strong pattern for failure occurrences.  In order to allow for more consistent test runs that better reflect the state of the Service Bus offering, this set of changes introduces infrastructure to allow Live tests an automatic retry on failure, to filter out tests that did not pass due to a timing issue.

## Goals
- Increase reliability of the tests during nightly runs, eliminating timing-related failures and adding stability to tests whose results are non-deterministic.

- Continue to adding polish to test infrastructure, ensuring that tests have a common base of utilities to work with, and which can clearly and easily be identified as such. 

- Attempt to minimize changes made, keeping the original approach, test cases, code styles

## Non-Goals
- Refactoring of the general approach for testing, the test cases, code style, and other artifacts of the codebase; changes should be kept as tightly scoped and as minimal as possible.

- Adding of additional test cases or new testing scenarios.

- Enabling test parallelism; _this will be a stretch goal in the second phase of these efforts._

## Areas of Discussion

In general, I'm not a fan of this approach of retying tests on failure, as I believe that it is a sign of unreliable tests that should be fixed.  However, in this case, we're dealing with an existing codebase of non-trivial size working against a service that is, by nature, non-deterministic.  I'm hoping that the pending work for phase 3 can help to unwind some of the timing issues, though I believe that we're nearing the end of changes with low effort and a high impact.  

As the Service Bus SDK receives future enhancements, I'm hopeful that we can continue to improve upon the tests as well to remove the reliance on timing;  at such time as we believe the tests are ready, the behavior for failure on retries can easily be toggled off with minimal effort needed.

### Infrastructure

- Created infrastructure to allow for tests tagged as "Live" to be automatically retried on failures by the test runner.  

- Combined attribution of Live tests with application of the appropriate test category, rather than necessitating two attributes.  

- Removed the former Live test annotation.  

### Live Tests

- Converted attribution of Live tests to take advantage of the new infrastructure, enabling automatic categorization and opting into retries on failure.  

- Temporarily disabling `QueueClientDiagnosticsTests.SendAndHandlerFireExceptionEvents` as it has been consistently unreliable during nightly runs and requires further investigation.

## Last Upstream Rebase

Friday, March 29, 2019 10:36am (EDT)

## Related and Follow-Up Issues

- [[Service Bus Client] Stabilize Live Tests](https://github.com/Azure/azure-sdk-for-net/issues/5446) (#5446)

### CODE REVIEW

File: src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Infrastructure/TestConstants.cs [hunk 1]
```diff
@@ -12,9 +12,13 @@ static class TestConstants
         
         // General 
         internal const string SessionPrefix = "session";
-        internal const int MaxAttemptsCount = 5;
+        internal const int MaxAttemptsCount = 5;
```
Comment (pakrym) [inline src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Infrastructure/TestConstants.cs]: nit: whitespace.

**Comment by pakrym** [comment]
Are all the live tests flaky? I'm not a fan of adding retry to everything because it might hide legit product issues. Can it be a property on Live attribute that controls if the test would be retried?

**Comment by jsquire** [comment]
> Are all the live tests flaky? I'm not a fan of adding retry to everything because it might hide legit product issues. Can it be a property on Live attribute that controls if the test would be retried?

I don't disagree.   (see the Areas of Discussion )    Unfortunately, there's no clear pattern for which tests are and are not.   We've been running them in a tight loop with each round of fixes and the failures are spread, other than two really angry tests.   If we spent time compiling results, I'm sure we could rule some out... but I can't justify the return on that time investment at this point.  When this SDK moves to track 2, I'm fully in favor of deprecating this approach.   Likewise, should I get to the point where  I feel that we've got enough stability without it, should that happen in the next round of changes where I focus on timing, likewise.

If it makes you feel any better, the max attempts for any given test can be set in the attribute;  setting to `1` will effectively give a test a single chance to pass with no retry.   As we gain confidence in any particular test, we can tune it down.

File: src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Infrastructure/TestConstants.cs (hunk 1 above)
Comment (jsquire) [reply src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Infrastructure/TestConstants.cs]: Thanks.    Worth calling out also, that this `MaxAttemptsCount` is not the "please retry a failed test this many times" control.   This was an existing value used in some tests to control the number of times that it attempts to receive messages that were sent and similar operations.

The `LiveTestDefaultMaxAttempts ` a few lines below is the test failure retry control.   It's currently set to just a single retry on failures to try and avoid hiding legitimate issues.

**Comment by mikeharder** [comment]
@jsquire: A few questions:

1. Do you believe it's impossible to write reliable tests given the current design of the client and service?  Or is just difficult rather than impossible, and it's not worth the effort to do so, particularly if this SDK will be rewritten for Track 2?

2. Can you share an example test (ideally the simplest example possible) showing the timing issues?

3. Do you have any sense how many failures are related to timing vs. transient failures?

**Comment by jsquire** [comment]
> @jsquire: A few questions:
> 
> 1. Do you believe it's impossible to write reliable tests given the current design of the client and service?  Or is just difficult rather than impossible, and it's not worth the effort to do so, particularly if this SDK will be rewritten for Track 2?
> 2. Can you share an example test (ideally the simplest example possible) showing the timing issues?
> 3. Do you have any sense how many failures are related to timing vs. transient failures?

I'm of the opinion that the tests are the culprit for the most part.  I suspect that it is possible to write a cleaner set of tests against the service using the SDK, but due to the non-deterministic nature of AMQP and messaging in general, it's much more difficult than it is with a request/response that is a more predictable flow.The tests overall have some design decisions that I find myself in disagreement with and I think that it shows they were written by folks with less experience, or at least those not overly familiar with asynchronous programming and messaging patterns.

**Comment by mikeharder** [comment]
In the most recent loop of 56 test runs, all tests were 100% reliable with the exception of `QueueClientDiagnosticsTests.SendAndHandlerFireExceptionEvents`.  Some other tests have failed occasionally, but the failure rate seems low enough that, in my opinion, I'd rather have occasional test failures than silent retries.

My preference would be to park this PR, fix `QueueClientDiagnosticsTests.SendAndHandlerFireExceptionEvents`, and see how reliable the tests are in nightly runs.  If the tests fail say one night every 2 months, we can live with that.

**Review by mikeharder — COMMENTED** [review]
In the most recent loop of 56 test runs, all tests were 100% reliable with the exception of `QueueClientDiagnosticsTests.SendAndHandlerFireExceptionEvents`.  Some other tests have failed occasionally, but the failure rate seems low enough that, in my opinion, I'd rather have occasional test failures than silent retries.

My preference would be to park this PR, fix `QueueClientDiagnosticsTests.SendAndHandlerFireExceptionEvents`, and see how reliable the tests are in nightly runs.  If the tests fail say one night every 2 months, we can live with that.

**Comment by jsquire** [comment]
> In the most recent loop of 56 test runs, all tests were 100% reliable with the exception of `QueueClientDiagnosticsTests.SendAndHandlerFireExceptionEvents`. Some other tests have failed occasionally, but the failure rate seems low enough that, in my opinion, I'd rather have occasional test failures than silent retries.
> 
> My preference would be to park this PR, fix `QueueClientDiagnosticsTests.SendAndHandlerFireExceptionEvents`, and see how reliable the tests are in nightly runs.  If the tests fail say one night every 2 months, we can live with that.

 I've no strong feeling about pushing these changes through; I think we all have the same concerns around the approach and don't advocate automatic retries as a good practice going forward.  

I'm happy to rebase the new work onto the phase 1 changes and leave these dangling for a bit in case we need them.  I'm going to leave the PR open until we get a chance to kick the tires on the next set of changes _(intended to shore up some of the timing and other misc things that I noted)_ and see where we want to go from there.

@AlexGhiondea:  Wanted to give you a chance to weigh in and make sure that you're comfortable with the proposed new approach.

**Comment by AlexGhiondea** [comment]
@jsquire what is the proposed plan? Leave the current PR without merging and see how stable the tests are?

**Comment by AlexGhiondea** [comment]
Sounds good -- let's try that and see how it goes!

**Comment by jsquire** [comment]
With the latest round of changes, I believe that I've gotten a solid baseline of stability and was able to further isolate.    As a result, I'm going to abandon this PR.   I'll keep the branch available should we see a need to revisit in the near future.   Thanks, all, for the feedback.

**Pull request closed without merging**